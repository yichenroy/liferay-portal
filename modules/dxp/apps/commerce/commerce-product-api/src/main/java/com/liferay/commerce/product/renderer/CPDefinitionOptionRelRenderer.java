/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 * <p>
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 * <p>
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.commerce.product.renderer;

import aQute.bnd.annotation.ProviderType;

import com.liferay.portal.kernel.exception.PortalException;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * @author Marco Leo
 */
@ProviderType
public interface CPDefinitionOptionRelRenderer {

	public String render(
		long cpDefinitionId, HttpServletRequest httpServletRequest,
		HttpServletResponse httpServletResponse) throws PortalException;

}