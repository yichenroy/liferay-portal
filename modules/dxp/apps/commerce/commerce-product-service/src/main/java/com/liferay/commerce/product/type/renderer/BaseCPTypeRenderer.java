/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.commerce.product.type.renderer;

import com.liferay.commerce.product.type.CPTypeRenderer;
import com.liferay.portal.kernel.log.Log;
import com.liferay.portal.kernel.log.LogFactoryUtil;
import com.liferay.portal.kernel.model.PortletConstants;
import com.liferay.portal.kernel.portlet.PortletBag;
import com.liferay.portal.kernel.portlet.PortletBagPool;
import com.liferay.portal.kernel.util.Portal;
import com.liferay.portal.kernel.util.Validator;
import com.liferay.portal.kernel.util.WebKeys;

import java.io.IOException;

import javax.servlet.RequestDispatcher;
import javax.servlet.ServletContext;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.osgi.service.component.annotations.Reference;

/**
 * @author Marco Leo
 */
public abstract class BaseCPTypeRenderer implements CPTypeRenderer {

	public void renderJSP(
			HttpServletRequest httpServletRequest,
			HttpServletResponse httpServletResponse, String path)
		throws IOException {

		renderJSP(
			getServletContext(httpServletRequest), httpServletRequest,
			httpServletResponse, path);
	}

	public void renderJSP(
			ServletContext servletContext,
			HttpServletRequest httpServletRequest,
			HttpServletResponse httpServletResponse, String path)
		throws IOException {

		RequestDispatcher requestDispatcher =
			servletContext.getRequestDispatcher(path);

		try {
			requestDispatcher.include(httpServletRequest, httpServletResponse);
		}
		catch (ServletException se) {
			_log.error("Unable to render JSP " + path, se);

			throw new IOException("Unable to render " + path, se);
		}
	}

	public void setServletContext(ServletContext servletContext) {
		_servletContext = servletContext;
	}

	protected ServletContext getServletContext(
		HttpServletRequest httpServletRequest) {

		if (_servletContext != null) {
			return _servletContext;
		}

		String portletId = portal.getPortletId(httpServletRequest);

		if (Validator.isNotNull(portletId)) {
			String rootPortletId = PortletConstants.getRootPortletId(portletId);

			PortletBag portletBag = PortletBagPool.get(rootPortletId);

			return portletBag.getServletContext();
		}

		return (ServletContext)httpServletRequest.getAttribute(WebKeys.CTX);
	}

	@Reference
	protected Portal portal;

	private static final Log _log = LogFactoryUtil.getLog(
		BaseCPTypeRenderer.class);

	private ServletContext _servletContext;

}