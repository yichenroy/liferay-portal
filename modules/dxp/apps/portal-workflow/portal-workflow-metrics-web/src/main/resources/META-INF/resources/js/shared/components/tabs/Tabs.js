import Icon from '../Icon';
import React from 'react';
import TabItem from './TabItem';
import {withRouter} from 'react-router-dom';

class Tabs extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			expanded: false
		};
	}

	hideNavbar() {
		this.setState({
			expanded: false
		});
	}

	toggleExpanded() {
		this.setState({
			expanded: !this.state.expanded
		});
	}

	render() {
		const {expanded} = this.state;
		const {
			location: {pathname, search},
			tabs
		} = this.props;

		const isActive = tab => pathname.includes(tab.key);

		const activeTab = tabs.filter(isActive)[0] || {};

		let navbarClassName = '';
		let togglerClassName = 'collapsed';

		if (expanded) {
			navbarClassName = 'show';
			togglerClassName = '';
		}

		return (
			<nav className='navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-secondary'>
				<div className='container-fluid container-fluid-max-xl'>
					<button
						aria-expanded={expanded}
						aria-label={Liferay.Language.get('toggle-navigation')}
						className={`${togglerClassName} navbar-toggler navbar-toggler-link`}
						onClick={this.toggleExpanded.bind(this)}
					>
						{activeTab.name}

						<Icon iconName='caret-bottom' />
					</button>

					<div
						className={`collapse navbar-collapse ${navbarClassName}`}
					>
						<div className='container-fluid container-fluid-max-xl'>
							<ul className='navbar-nav'>
								{tabs.map((tab, index) => (
									<li
										className='nav-item'
										key={index}
										onClick={this.hideNavbar.bind(this)}
									>
										<TabItem
											{...tab}
											active={isActive(tab)}
											query={search}
										/>
									</li>
								))}
							</ul>
						</div>
					</div>
				</div>
			</nav>
		);
	}
}

export default withRouter(Tabs);
export {Tabs};
