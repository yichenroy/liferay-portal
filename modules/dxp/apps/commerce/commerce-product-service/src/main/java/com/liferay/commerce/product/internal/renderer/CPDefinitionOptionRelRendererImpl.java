/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.commerce.product.internal.renderer;

import com.liferay.commerce.product.model.CPDefinitionOptionRel;
import com.liferay.commerce.product.model.CPDefinitionOptionValueRel;
import com.liferay.commerce.product.renderer.CPDefinitionOptionRelRenderer;
import com.liferay.commerce.product.service.CPDefinitionOptionRelService;
import com.liferay.dynamic.data.mapping.form.renderer.DDMFormRenderer;
import com.liferay.dynamic.data.mapping.form.renderer.DDMFormRenderingContext;
import com.liferay.dynamic.data.mapping.model.DDMForm;
import com.liferay.dynamic.data.mapping.model.DDMFormField;
import com.liferay.dynamic.data.mapping.model.DDMFormFieldOptions;
import com.liferay.dynamic.data.mapping.model.LocalizedValue;
import com.liferay.portal.kernel.exception.PortalException;
import com.liferay.portal.kernel.util.Portal;

import java.util.List;
import java.util.Locale;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.osgi.service.component.annotations.Component;
import org.osgi.service.component.annotations.Reference;

/**
 * @author Marco Leo
 */
@Component(immediate = true)
public class CPDefinitionOptionRelRendererImpl
	implements CPDefinitionOptionRelRenderer {

	@Override
	public String render(
			long cpDefinitionId, HttpServletRequest httpServletRequest,
			HttpServletResponse httpServletResponse)
		throws PortalException {

		DDMForm ddmForm = new DDMForm();

		Locale locale = _portal.getLocale(httpServletRequest);

		List<CPDefinitionOptionRel> cpDefinitionOptionRels =
			_cpDefinitionOptionRelService.
				getSkuContributorCPDefinitionOptionRels(cpDefinitionId);

		for (CPDefinitionOptionRel cpDefinitionOptionRel :
				cpDefinitionOptionRels) {

			List<CPDefinitionOptionValueRel> cpDefinitionOptionValueRels =
				cpDefinitionOptionRel.getCPDefinitionOptionValueRels();

			DDMFormField ddmFormField = new DDMFormField(
				String.valueOf(cpDefinitionOptionRel.
					getCPDefinitionOptionRelId()),
				cpDefinitionOptionRel.getDDMFormFieldTypeName());

			if (!cpDefinitionOptionValueRels.isEmpty()) {
				DDMFormFieldOptions ddmFormFieldOptions =
					new DDMFormFieldOptions();

				for (CPDefinitionOptionValueRel cpDefinitionOptionValueRel :
						cpDefinitionOptionValueRels) {

					ddmFormFieldOptions.addOptionLabel(
						String.valueOf(cpDefinitionOptionValueRel.
							getCPDefinitionOptionValueRelId()),
						locale, cpDefinitionOptionValueRel.getTitle(locale));
				}

				ddmFormField.setDDMFormFieldOptions(ddmFormFieldOptions);
			}

			LocalizedValue localizedValue = new LocalizedValue(locale);

			localizedValue.addString(
				locale, cpDefinitionOptionRel.getName(locale));

			ddmFormField.setLabel(localizedValue);

			ddmFormField.setRequired(true);

			ddmForm.addDDMFormField(ddmFormField);
		}

		DDMFormRenderingContext ddmFormRenderingContext =
			new DDMFormRenderingContext();

		ddmFormRenderingContext.setContainerId("options");
		ddmFormRenderingContext.setHttpServletRequest(httpServletRequest);
		ddmFormRenderingContext.setHttpServletResponse(httpServletResponse);
		ddmFormRenderingContext.setLocale(locale);

		return _ddmFormRenderer.render(ddmForm, ddmFormRenderingContext);
	}

	@Reference
	private CPDefinitionOptionRelService _cpDefinitionOptionRelService;

	@Reference
	private DDMFormRenderer _ddmFormRenderer;

	@Reference
	private Portal _portal;

}