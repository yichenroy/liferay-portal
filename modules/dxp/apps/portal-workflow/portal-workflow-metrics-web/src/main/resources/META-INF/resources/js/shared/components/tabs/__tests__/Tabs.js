import React from 'react';
import renderer from 'react-test-renderer';
import { MockRouter as Router } from '../../../../test/mock/MockRouter';
import { Tabs } from '../Tabs';

const tabs = [
	{
		key: 'completed',
		name: Liferay.Language.get('completed'),
		params: {
			processId: 35135
		},
		path: '/dashboard/:processId/completed'
	},
	{
		key: 'pending',
		name: Liferay.Language.get('pending'),
		params: {
			processId: 35135
		},
		path: '/dashboard/:processId/pending'
	}
];

test('Should expand tab items', () => {
	const component = mount(
		<Router>
			<Tabs
				location={{
					pathname: '/dashboard/:processId/completed'
				}}
				tabs={tabs}
			/>
		</Router>
	);

	const instance = component.find(Tabs).instance();

	instance.toggleExpanded();

	expect(component).toMatchSnapshot();
});

test('Should hide tab items', () => {
	const component = mount(
		<Router>
			<Tabs
				location={{
					pathname: '/dashboard/:processId/completed'
				}}
				tabs={tabs}
			/>
		</Router>
	);

	const instance = component.find(Tabs).instance();

	instance.toggleExpanded();
	instance.hideNavbar();

	expect(component).toMatchSnapshot();
});

test('Should render component', () => {
	const component = renderer.create(
		<Router>
			<Tabs
				location={{
					pathname: '/dashboard/:processId/pending'
				}}
				tabs={tabs}
			/>
		</Router>
	);

	const tree = component.toJSON();

	expect(tree).toMatchSnapshot();
});