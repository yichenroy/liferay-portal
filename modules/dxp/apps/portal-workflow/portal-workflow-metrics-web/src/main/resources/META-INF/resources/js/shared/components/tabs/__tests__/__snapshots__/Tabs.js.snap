// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should expand tab items 1`] = `
<MockRouter>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "match": Object {
            "params": Object {
              "page": 1,
              "search": undefined,
              "sort": undefined,
            },
          },
          "pathname": "/processes/1/10/title%3Aasc",
          "search": "?backPath=%2F",
        },
      ]
    }
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "match": Object {
                "params": Object {
                  "page": 1,
                  "search": undefined,
                  "sort": undefined,
                },
              },
              "pathname": "/processes/1/10/title%3Aasc",
              "search": "?backPath=%2F",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "match": Object {
              "params": Object {
                "page": 1,
                "search": undefined,
                "sort": undefined,
              },
            },
            "pathname": "/processes/1/10/title%3Aasc",
            "search": "?backPath=%2F",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Tabs
        location={
          Object {
            "pathname": "/dashboard/:processId/completed",
          }
        }
        tabs={
          Array [
            Object {
              "key": "completed",
              "name": "completed",
              "params": Object {
                "processId": 35135,
              },
              "path": "/dashboard/:processId/completed",
            },
            Object {
              "key": "pending",
              "name": "pending",
              "params": Object {
                "processId": 35135,
              },
              "path": "/dashboard/:processId/pending",
            },
          ]
        }
      >
        <nav
          className="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-secondary"
        >
          <div
            className="container-fluid container-fluid-max-xl"
          >
            <button
              aria-expanded={false}
              aria-label="toggle-navigation"
              className="collapsed navbar-toggler navbar-toggler-link"
              onClick={[Function]}
            >
              completed
              <Icon
                iconName="caret-bottom"
              >
                <svg
                  className="lexicon-icon lexicon-icon-caret-bottom"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "<use xlink:href=\\"http://localhost:8080/o/admin-theme/images/lexicon/icons.svg#caret-bottom\\" />",
                    }
                  }
                  focusable="false"
                  role="presentation"
                />
              </Icon>
            </button>
            <div
              className="collapse navbar-collapse "
            >
              <div
                className="container-fluid container-fluid-max-xl"
              >
                <ul
                  className="navbar-nav"
                >
                  <li
                    className="nav-item"
                    key="0"
                    onClick={[Function]}
                  >
                    <TabItem
                      active={true}
                      key="completed"
                      name="completed"
                      params={
                        Object {
                          "processId": 35135,
                        }
                      }
                      path="/dashboard/:processId/completed"
                    >
                      <Link
                        aria-label="current-page"
                        className="active nav-link"
                        to={
                          Object {
                            "pathname": "/dashboard/35135/completed",
                            "search": undefined,
                          }
                        }
                      >
                        <a
                          aria-label="current-page"
                          className="active nav-link"
                          href="/dashboard/35135/completed"
                          onClick={[Function]}
                        >
                          <span
                            className="navbar-text-truncate"
                          >
                            completed
                          </span>
                        </a>
                      </Link>
                    </TabItem>
                  </li>
                  <li
                    className="nav-item"
                    key="1"
                    onClick={[Function]}
                  >
                    <TabItem
                      active={false}
                      key="pending"
                      name="pending"
                      params={
                        Object {
                          "processId": 35135,
                        }
                      }
                      path="/dashboard/:processId/pending"
                    >
                      <Link
                        aria-label=""
                        className=" nav-link"
                        to={
                          Object {
                            "pathname": "/dashboard/35135/pending",
                            "search": undefined,
                          }
                        }
                      >
                        <a
                          aria-label=""
                          className=" nav-link"
                          href="/dashboard/35135/pending"
                          onClick={[Function]}
                        >
                          <span
                            className="navbar-text-truncate"
                          >
                            pending
                          </span>
                        </a>
                      </Link>
                    </TabItem>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </nav>
      </Tabs>
    </Router>
  </MemoryRouter>
</MockRouter>
`;

exports[`Should hide tab items 1`] = `
<MockRouter>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "match": Object {
            "params": Object {
              "page": 1,
              "search": undefined,
              "sort": undefined,
            },
          },
          "pathname": "/processes/1/10/title%3Aasc",
          "search": "?backPath=%2F",
        },
      ]
    }
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "match": Object {
                "params": Object {
                  "page": 1,
                  "search": undefined,
                  "sort": undefined,
                },
              },
              "pathname": "/processes/1/10/title%3Aasc",
              "search": "?backPath=%2F",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "match": Object {
              "params": Object {
                "page": 1,
                "search": undefined,
                "sort": undefined,
              },
            },
            "pathname": "/processes/1/10/title%3Aasc",
            "search": "?backPath=%2F",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Tabs
        location={
          Object {
            "pathname": "/dashboard/:processId/completed",
          }
        }
        tabs={
          Array [
            Object {
              "key": "completed",
              "name": "completed",
              "params": Object {
                "processId": 35135,
              },
              "path": "/dashboard/:processId/completed",
            },
            Object {
              "key": "pending",
              "name": "pending",
              "params": Object {
                "processId": 35135,
              },
              "path": "/dashboard/:processId/pending",
            },
          ]
        }
      >
        <nav
          className="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-secondary"
        >
          <div
            className="container-fluid container-fluid-max-xl"
          >
            <button
              aria-expanded={false}
              aria-label="toggle-navigation"
              className="collapsed navbar-toggler navbar-toggler-link"
              onClick={[Function]}
            >
              completed
              <Icon
                iconName="caret-bottom"
              >
                <svg
                  className="lexicon-icon lexicon-icon-caret-bottom"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "<use xlink:href=\\"http://localhost:8080/o/admin-theme/images/lexicon/icons.svg#caret-bottom\\" />",
                    }
                  }
                  focusable="false"
                  role="presentation"
                />
              </Icon>
            </button>
            <div
              className="collapse navbar-collapse "
            >
              <div
                className="container-fluid container-fluid-max-xl"
              >
                <ul
                  className="navbar-nav"
                >
                  <li
                    className="nav-item"
                    key="0"
                    onClick={[Function]}
                  >
                    <TabItem
                      active={true}
                      key="completed"
                      name="completed"
                      params={
                        Object {
                          "processId": 35135,
                        }
                      }
                      path="/dashboard/:processId/completed"
                    >
                      <Link
                        aria-label="current-page"
                        className="active nav-link"
                        to={
                          Object {
                            "pathname": "/dashboard/35135/completed",
                            "search": undefined,
                          }
                        }
                      >
                        <a
                          aria-label="current-page"
                          className="active nav-link"
                          href="/dashboard/35135/completed"
                          onClick={[Function]}
                        >
                          <span
                            className="navbar-text-truncate"
                          >
                            completed
                          </span>
                        </a>
                      </Link>
                    </TabItem>
                  </li>
                  <li
                    className="nav-item"
                    key="1"
                    onClick={[Function]}
                  >
                    <TabItem
                      active={false}
                      key="pending"
                      name="pending"
                      params={
                        Object {
                          "processId": 35135,
                        }
                      }
                      path="/dashboard/:processId/pending"
                    >
                      <Link
                        aria-label=""
                        className=" nav-link"
                        to={
                          Object {
                            "pathname": "/dashboard/35135/pending",
                            "search": undefined,
                          }
                        }
                      >
                        <a
                          aria-label=""
                          className=" nav-link"
                          href="/dashboard/35135/pending"
                          onClick={[Function]}
                        >
                          <span
                            className="navbar-text-truncate"
                          >
                            pending
                          </span>
                        </a>
                      </Link>
                    </TabItem>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </nav>
      </Tabs>
    </Router>
  </MemoryRouter>
</MockRouter>
`;

exports[`Should render component 1`] = `
<nav
  className="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-secondary"
>
  <div
    className="container-fluid container-fluid-max-xl"
  >
    <button
      aria-expanded={false}
      aria-label="toggle-navigation"
      className="collapsed navbar-toggler navbar-toggler-link"
      onClick={[Function]}
    >
      pending
      <svg
        className="lexicon-icon lexicon-icon-caret-bottom"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<use xlink:href=\\"http://localhost:8080/o/admin-theme/images/lexicon/icons.svg#caret-bottom\\" />",
          }
        }
        focusable="false"
        role="presentation"
      />
    </button>
    <div
      className="collapse navbar-collapse "
    >
      <div
        className="container-fluid container-fluid-max-xl"
      >
        <ul
          className="navbar-nav"
        >
          <li
            className="nav-item"
            onClick={[Function]}
          >
            <a
              aria-label=""
              className=" nav-link"
              href="/dashboard/35135/completed"
              onClick={[Function]}
            >
              <span
                className="navbar-text-truncate"
              >
                completed
              </span>
            </a>
          </li>
          <li
            className="nav-item"
            onClick={[Function]}
          >
            <a
              aria-label="current-page"
              className="active nav-link"
              href="/dashboard/35135/pending"
              onClick={[Function]}
            >
              <span
                className="navbar-text-truncate"
              >
                pending
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;
